cmd/server/main.go

```diff
@@ -18,15 +25,58 @@ func main() {
 		log.Fatal("failed to provide server config: " + err.Error())
 	}
 
-	run(appConfig)
+	if err = run(appConfig); err != nil {
+		appConfig.Log.ErrorContext(context.Background(), "failed to run the server: "+err.Error())
+		os.Exit(1)
+	}
 }
 
-func run(cfg *config.ServerConfig) {
-	store := storage.NewMemStorage()
+func run(cfg *config.ServerConfig) error {
+	cfg.Log.DebugContext(context.Background(), "provided config",
+		slog.String("ServerAddress", cfg.Address),
+		slog.String("StorePath", cfg.StorePath),
+		slog.String("LogLevel", cfg.LogLevel),
+		slog.Int("StoreInterval", cfg.StoreInterval),
+		slog.Bool("ReStore", cfg.ReStore))
 
+	store, err := storage.NewStorage(cfg)
+	if err != nil {
+		cfg.Log.ErrorContext(context.Background(), "failed to initialize the mem")
+		return fmt.Errorf("failed to setup the memstore: %w", err)
+	}
 	repos := repositories.NewRepository(store)
-	h := handlers.NewHandler(server.NewService(repos, cfg.Log), cfg.Log)
+	h := handlers.NewHandler(server.NewMetricsService(repos, cfg.Log), cfg.Log)
 
 	cfg.Log.DebugContext(context.Background(), "running server 🔥 on port: "+cfg.Address)
-	log.Fatal(http.ListenAndServe(cfg.Address, h.InitHandlers()))
+	srv := &http.Server{
+		Addr:    cfg.Address,
+		Handler: h.InitHandlers(),
+	}
+
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
```

Comment:
I am not going to check the graceful shutdown mechanism now, it is due in the 4th sprint, I'll review it then